tailscale/client.go

```diff
@@ -937,6 +937,160 @@ func (c *Client) SetDeviceIPv4Address(ctx context.Context, deviceID string, ipv4
 	return c.performRequest(req, nil)
 }
 
+const (
+	EmptyProviderType      ProviderType = ""
+	SlackProviderType      ProviderType = "slack"
+	MattermostProviderType ProviderType = "mattermost"
+	GoogleChatProviderType ProviderType = "googlechat"
+	DiscordProviderType    ProviderType = "discord"
+)
+
+const (
+	NodeCreated                    SubscriptionType = "nodeCreated"
+	NodeNeedsApproval              SubscriptionType = "nodeNeedsApproval"
+	NodeApproved                   SubscriptionType = "nodeApproved"
+	NodeKeyExpiringInOneDay        SubscriptionType = "nodeKeyExpiringInOneDay"
+	NodeKeyExpired                 SubscriptionType = "nodeKeyExpired"
+	NodeDeleted                    SubscriptionType = "nodeDeleted"
+	PolicyUpdate                   SubscriptionType = "policyUpdate"
+	UserCreated                    SubscriptionType = "userCreated"
+	UserNeedsApproval              SubscriptionType = "userNeedsApproval"
+	UserSuspended                  SubscriptionType = "userSuspended"
+	UserRestored                   SubscriptionType = "userRestored"
+	UserDeleted                    SubscriptionType = "userDeleted"
+	UserApproved                   SubscriptionType = "userApproved"
+	UserRoleUpdated                SubscriptionType = "userRoleUpdated"
+	SubnetIPForwardingNotEnabled   SubscriptionType = "subnetIPForwardingNotEnabled"
+	ExitNodeIPForwardingNotEnabled SubscriptionType = "exitNodeIPForwardingNotEnabled"
+)
+
+type (
+	// ProviderType defines the provider type for a Webhook destination.
+	ProviderType string
+
+	// SubscriptionType defines events in tailscale to subscribe a Webhook to.
+	SubscriptionType string
+
+	// Webhook type defines a webhook endpoint within a tailnet.
+	Webhook struct {
+		EndpointID       string             `json:"endpointId"`
+		EndpointURL      string             `json:"endpointUrl"`
+		ProviderType     ProviderType       `json:"providerType"`
+		CreatorLoginName string             `json:"creatorLoginName"`
+		Created          time.Time          `json:"created"`
+		LastModified     time.Time          `json:"lastModified"`
+		Subscriptions    []SubscriptionType `json:"subscriptions"`
+		Secret           string             `json:"secret,omitempty"`
+	}
+
+	// CreateWebhookRequest type describes the configuration for creating a Webhook.
+	CreateWebhookRequest struct {
+		EndpointURL   string             `json:"endpointUrl"`
+		ProviderType  ProviderType       `json:"providerType"`
+		Subscriptions []SubscriptionType `json:"subscriptions"`
+	}
+)
+
+// CreateWebhook creates a new webhook with the specifications provided in the CreateWebhookRequest.
+// Returns a Webhook if successful.
+func (c *Client) CreateWebhook(ctx context.Context, request CreateWebhookRequest) (Webhook, error) {
+	const uriFmt = "/api/v2/tailnet/%s/webhooks"
+
+	req, err := c.buildRequest(ctx, http.MethodPost, fmt.Sprintf(uriFmt, c.tailnet), requestBody(request))
```

Comment:
Need to URL escape `c.tailnet` here and in other places where we build the URL. Might be worth factoring into a `tailnetURL(tmpl)` function or something.

Reply:
Good call, will look at doing this as a follow up PR.

Reply:
Why not do it in this PR, at least for the new functions?

Reply:
Was thinking it would be good to do as a single sweep / dedicated PR since it touches more than the webhook endpoints, can look into adding this for the new functions here though.

Reply:
Ended up storing this in a `tailnetPathEscaped` field on the struct and replacing existing usages of `tailnet` in path segments with this field instead. Left the original `tailnet` field for now in the case that we want / need the un-escaped version for some reason.